### ISSUE TITLE
Gpbackup cannot fully support gpcopy in GPDB7

### ISSUE THREAD

**Issue by xiaoxiaoHe-E** [greenplum-db/gpbackup#614]
In current gpcopy, users can specify the dumper by `--dumper gpbackup` in gpcopy command to use gpbackup to dump DDL.  But it's still a Beta feature since there are some known issues with gpbackup. 

We hope the team can fix these issues, so maybe gpcopy can use gpdbackup as the default dumper in GPDB7.

Below is what we found when integrating gpbackup as a Beta feature for gpcopy in 2020.  Maybe some of the issues are fixed by now. 

We will update the issues when gpcopy test pipeline in GPDB7 is ready.

> The relevant PR can be found at https://github.com/pivotal/gpcopy/pull/250
> However, there are a few issues we found when making the integration:
> - gpbackup always creates a lock file with the current timestamp in tmp directory. It prevents gpcopy to dump DDL in parallel (at a second). By checking the code, I found the lock actually didn't matter to gpcopy -- gpcopy always specify the tables to dump and a unique output directory. So we work around it by just deleting the lock file when gpbackup complains. Maybe one possible solution in gpbackup side is to put the lock file into the specified output directory? The relevant code https://github.com/greenplum-db/gpbackup/blob/0854b55c23210fabff39572755a27373bf0b241c/backup/wrappers.go#L128
> - '--no-owner' and '--no-privileges' options which pg_dump supports. These are quite useful to gpcopy if the destination doesn't have the relevant roles. Currently, gpcopy has to fallback to pg_dump if the required roles don't exist.
> - gpbackup cannot handle dot '.' in the '--include-table' option.  This is a minor issue to gpcopy. Currently, gpcopy will try to detect if there is a '.' in the table/schema name, then fallback to pg_dump if needed. See: https://github.com/greenplum-db/gpbackup/blob/35e8262e59adb02ebe67c63ec6fb79272c12ce77/options/options.go#L175
> - gpbackup always dump sequence value even with '--metadata-only' option. The pg_dump for GPDB4 and GPDB5 actually have the same behavior. But it was changed in GPDB6. When the '--metadata-only' is given to the pg_dump, the value of sequences won't be dumped. So maybe gpbackup should align the behavior with the latest pg_dump?
> - It seems that gpbackup doesn't support kerberos authentication. gpcopy used to have the same problem which has been fixed by this PR https://github.com/pivotal/gpcopy/pull/140 .
> Right now, the lock file and privilege issues are quite annoyed to gpcopy. Appreciated if you can take a look :)

**Comment by khuddlefish** [comment]
Hi @xiaoxiaoHe-E ,

On the lockfile topic, gpbackup creates that because we use timestamps as a key for all of our backup files and so having mutilple gpbackup processes running at the same time would cause file collisions. For customers that want to do parallel backups of different sets, we usually recommend scripting to sleep for 1 second between starting each gpbackup process so they will all have unique timestamps. Is that something that you think might work for gpcopy? 1 second is not nothing, but it shouldn't add much overhead as long as you are not doing too many processes.

Otherwise, we wanted to understand more about your process for running in parallel. How do you batch your objects? How are you planning to call gpbackup?

I'm not sure I understand what you mean by the solution of putting the gpbackup lockfile in the specified directory. Do you mean checking if there is one with the same timestamp in the same directory (specififed backup directory) but allowing the same timestamp if it's a different directory? I think that might cause problems for backups that include data and use the gpbackup_helper because we create pipe files to send data that are named with the segment and timestamp, but not named with the backup directory. Looking at the code we may also include the PID in those names, so we could investigate that option and see if it's possible.


For the other asks, it would be helpful to us if you could give us a sense of priority for them since there are several items and most probably aren't trivial for us to support. Are any of these blockers for gpbackup being the default dumper for pg_dump?  Are some of them easier to workaround than others? I remember in the other issue you mentioned an issue where gpbackup wasn't able to backup a table partitioned by a type, is that still relevant? If you can let us know your top 3 priority items, we can investigate them for you and see how difficult it would be to support your asks.

**Comment by xiaoxiaoHe-E** [comment]
Hi, I find a new issue of gpbackup: when using gpbackup to dump DDL for table A, it will also dump table B that inherits table A.
create table:
```
CREATE TABLE foo(a int);
CREATE TABLE too(b int);
CREATE TABLE footoo() INHERITS(foo,too);
```
gpbakcup command:
```
$ gpbackup --version
gpbackup version 1.27.0+dev.12.g0a53f73

gpbackup --dbname testdb_gpcopy_ginkgo1 --backup-dir /tmp/gpcopy_dump_JBMugKXNagLqRMWSno6o58 --metadata-only --include-table "public"."too"
```
output of gpbackup:
```
SET client_encoding = 'UTF8';


CREATE TABLE public.too (
	b integer
) DISTRIBUTED BY (b);


ALTER TABLE public.too OWNER TO gpadmin;


CREATE TABLE public.footoo (
	a integer,
	b integer
) INHERITS (public.foo, public.too) DISTRIBUTED BY (a);


ALTER TABLE public.footoo OWNER TO gpadmin;
```
gpdb version: `Greenplum Database 7.0.0-beta.1+dev.167.g4d7c52a5ae build dev`

when restore the output SQL, there will be an error:
```
 pq: relation "public.foo" does not exist
```

**Comment by xiaoxiaoHe-E** [comment]
for tables:
```sql
CREATE TABLE public.cities (
	name text,
	altitude integer
) DISTRIBUTED BY (name);

CREATE TABLE inherit_test.capitals (
	name text,
	altitude integer,
	state character(2)
) INHERITS (public.cities) DISTRIBUTED BY (name);

CREATE TABLE inherit_test.important_capitals (
	name text,
	altitude integer,
	state character(2),
	descr character(2)
) INHERITS (inherit_test.capitals) DISTRIBUTED BY (name);
```
gpbackup also has different behavior in gpdb7 and gpdb6
gpbackup command:
```
$ gpbackup --version
gpbackup version 1.27.0+dev.12.g0a53f73

gpbackup --dbname testdb_gpcopy_ginkgo1 --backup-dir /tmp/gpcopy_dump_wbmJbrAmnSvuGMwvbdF3k5 --metadata-only --include-table "inherit_test"."capitals"
```
We wish gpbackup should only dump the `inherit_test.capitals`

**Comment by xiaoxiaoHe-E** [comment]
Sorry for replying late, I didn't see the comment.
We are working on gpcopy to support GPDB7. pg_dump and gpbackup are used as DDL dumper by gpcopy. Gpcopy uses gpbackup or pgdump to dump and restore only the DDL. And then copy the data by copy sql command.

On the lockfile topic, gpcopy only uses gpbackup to copy the DDL, not include the data. The data integrity doesn't matter to gpcopy. Can you add another value like task number to avoid file collisions? 

In a gpcopy job, we will 
1. decide which tables to copy and the order to copy according to the dependence between the tables
3. generate the copy task graph, one task copy one table (include DDL and data)
4. in one task, gpcopy uses gpbackup or pgdump to dump and restore only the DDL. And then copy the data by copy sql command.
5. start workers to work on ready tasks until tasks are all done

For other issues, Kerberos and `gpbackup cannot handle dot '.' ` is less of priority. We hope others can be fixed first. Thanks for your help :)

**Comment by xiaoxiaoHe-E** [comment]
Hi, team:
I have finished the gpcopy test cases in GPDB7 just now. Above are all the issues we found for gpbackup. Gpcopy in GPDB7 now uses gpbackup as the default DDL dumper now. For the above issues, we can handle them by falling back to using pg_dump as the DDL dumper.
Still, we hope you can fix these known issues so that at least, we can have one utility that can completely support gpcopy in GPDB7. It's kind of tricky to switch between gpbackup and pg_dump. Thanks for your help.

**Comment by khuddlefish** [comment]
Hi @xiaoxiaoHe-E, we've made some progress on some of the requests from gpcopy and want to check in again.

What we've delivered:
- If gpcopy runs with --backup-dir, --metadata-only and --no-history gpbackup will write the lockfile to the backup-dir and allow multiple concurrent backups to run as long as distinct backup-dirs are used.
- Some fixes to inheritance behavior.

Things remaining:
- --no-inherits behavior does not match pg_dump. We have looked at this and feel strongly that we do not want gpbackup to dump a table without the parent it inherits from. We don't think it would support most of our users use cases and would cause more problems than it helps. If it really causes an issue for your team then we might be willing to talk more about how to resolve it.
- --no-owner and --no-privileges options. I don't think we've gotten this request from other customers but we can see how it might be valuable and we feel we could probably accomplish it.
- gpbackup cannot handle '.' in the table name. We think we could probably fix this if needed.
- gpbackup always dumps sequence value and behavior differs from pg_dump. We looked at this and we do consider the initial sequence value to be metadata so we do not intend to change this behavior.
- gpbackup doesn't support kerberos authentication. Our impression is that this should be something fully handled in the user's environment and that we shouldn't need to make changes to gpbackup to support kerberos authentication. Do you know of actual cases where it hasn't worked? We'd be open to talking about this more.

I do want to make clear that it is not and has never been a goal for gpbackup to have feature parity with pg_dump. They are different products and can be used in different ways. If we add a new feature in gpbackup we want a real use case where it would benefit a customer, not just because that's the way pg_dump works.

For the remaining features can you give us more info about how high priority they would be for you team? Are they common scenarios in user workflows or rare edge cases? Does it complicate the gpcopy code a lot to have to handle the different cases? Also did gpcopy switch to use gpbackup as the default dumper or is it still using pg_dump? If it hasn't switched what more would need to happen before your team makes the switch? From our side, we have still not gotten clear messaging about how important these features are to gpcopy so we have been hesitant to implement them without a compelling reason. Let's have another meeting to discuss.
